Flaky TestHivePushdownFilterQueries

testFilterFunctions and testArrays failed  with the following stacktrace in a recent build: https://api.travis-ci.org/v3/job/572454390/log.txt

```
[ERROR] testFilterFunctions(com.facebook.presto.hive.TestHivePushdownFilterQueries)  Time elapsed: 12.497 s  <<< FAILURE!
java.lang.AssertionError: Execution of 'actual' query failed: SELECT * FROM lineitem_ex WHERE keys[1] % 5 = 0 AND nested_keys[1][2] > 100
	at org.testng.Assert.fail(Assert.java:83)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:150)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:138)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.assertQueryUsingH2Cte(TestHivePushdownFilterQueries.java:337)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.assertFilterProject(TestHivePushdownFilterQueries.java:235)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.testFilterFunctions(TestHivePushdownFilterQueries.java:277)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Failed to read ORC file: file:/tmp/PrestoTest3034274899939071963/hive_data/tpch/lineitem_ex/20190815_205249_00016_kkt5v_c501edb6-9447-4d8a-8374-2f5c02842425
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:115)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:391)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:147)
	... 18 more
Caused by: com.facebook.presto.spi.PrestoException: Failed to read ORC file: file:/tmp/PrestoTest3034274899939071963/hive_data/tpch/lineitem_ex/20190815_205249_00016_kkt5v_c501edb6-9447-4d8a-8374-2f5c02842425
	at com.facebook.presto.hive.orc.OrcSelectivePageSource.getNextPage(OrcSelectivePageSource.java:94)
	at com.facebook.presto.operator.TableScanOperator.getOutput(TableScanOperator.java:250)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:379)
	at com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:283)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:675)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:276)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1077)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:483)
	at com.facebook.presto.$gen.Presto_0_225_SNAPSHOT_4222738__testversion____20190815_205238_14609.run(Unknown Source)
	... 3 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
	at com.facebook.presto.orc.reader.ListFilter.populateElementFilters(ListFilter.java:213)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.readNotAllNulls(ListSelectiveStreamReader.java:369)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.read(ListSelectiveStreamReader.java:226)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.readNotAllNulls(ListSelectiveStreamReader.java:373)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.read(ListSelectiveStreamReader.java:226)
	at com.facebook.presto.orc.OrcSelectiveRecordReader.getNextPage(OrcSelectiveRecordReader.java:263)
	at com.facebook.presto.hive.orc.OrcSelectivePageSource.getNextPage(OrcSelectivePageSource.java:78)
	... 12 more

[ERROR] testArrays(com.facebook.presto.hive.TestHivePushdownFilterQueries)  Time elapsed: 2.158 s  <<< FAILURE!
java.lang.AssertionError: Execution of 'actual' query failed: SELECT * FROM lineitem_ex WHERE nested_keys[3] IS NOT NULL
	at org.testng.Assert.fail(Assert.java:83)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:150)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:138)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.assertQueryUsingH2Cte(TestHivePushdownFilterQueries.java:337)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.assertFilterProject(TestHivePushdownFilterQueries.java:235)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.testArrays(TestHivePushdownFilterQueries.java:188)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Failed to read ORC file: file:/tmp/PrestoTest3034274899939071963/hive_data/tpch/lineitem_ex/20190815_205249_00016_kkt5v_c501edb6-9447-4d8a-8374-2f5c02842425
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:115)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:391)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:147)
	... 18 more
Caused by: com.facebook.presto.spi.PrestoException: Failed to read ORC file: file:/tmp/PrestoTest3034274899939071963/hive_data/tpch/lineitem_ex/20190815_205249_00016_kkt5v_c501edb6-9447-4d8a-8374-2f5c02842425
	at com.facebook.presto.hive.orc.OrcSelectivePageSource.getNextPage(OrcSelectivePageSource.java:94)
	at com.facebook.presto.operator.TableScanOperator.getOutput(TableScanOperator.java:250)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:379)
	at com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:283)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:675)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:276)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1077)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:483)
	at com.facebook.presto.$gen.Presto_0_225_SNAPSHOT_4222738__testversion____20190815_205237_14608.run(Unknown Source)
	... 3 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
	at com.facebook.presto.orc.reader.ListFilter.populateElementFilters(ListFilter.java:213)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.readNotAllNulls(ListSelectiveStreamReader.java:369)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.read(ListSelectiveStreamReader.java:226)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.readNotAllNulls(ListSelectiveStreamReader.java:373)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.read(ListSelectiveStreamReader.java:226)
	at com.facebook.presto.orc.OrcSelectiveRecordReader.getNextPage(OrcSelectiveRecordReader.java:263)
	at com.facebook.presto.hive.orc.OrcSelectivePageSource.getNextPage(OrcSelectivePageSource.java:78)
	... 12 more
```

@mbasmanova based on the stack trace it looks related to https://github.com/prestodb/presto/pull/13099.  Can you take a look?

Some more flakiness:

```
ERROR] Tests run: 1449, Failures: 1, Errors: 0, Skipped: 57, Time elapsed: 2,528.97 s <<< FAILURE! - in TestSuite
[ERROR] testArrays(com.facebook.presto.hive.TestHivePushdownFilterQueries)  Time elapsed: 36.627 s  <<< FAILURE!
java.lang.AssertionError: Execution of 'actual' query failed: SELECT * FROM lineitem_ex WHERE nested_keys[3] IS NULL
	at org.testng.Assert.fail(Assert.java:83)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:150)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:138)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.assertQueryUsingH2Cte(TestHivePushdownFilterQueries.java:366)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.assertFilterProject(TestHivePushdownFilterQueries.java:264)
	at com.facebook.presto.hive.TestHivePushdownFilterQueries.testArrays(TestHivePushdownFilterQueries.java:197)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Failed to read ORC file: file:/tmp/PrestoTest8169961107556799537/hive_data/tpch/lineitem_ex/20190820_080456_00016_qsgeg_149acf7a-5e90-4542-883f-0e6d386167c9
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:115)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:391)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:147)
	... 18 more
Caused by: com.facebook.presto.spi.PrestoException: Failed to read ORC file: file:/tmp/PrestoTest8169961107556799537/hive_data/tpch/lineitem_ex/20190820_080456_00016_qsgeg_149acf7a-5e90-4542-883f-0e6d386167c9
	at com.facebook.presto.hive.orc.OrcSelectivePageSource.getNextPage(OrcSelectivePageSource.java:94)
	at com.facebook.presto.operator.TableScanOperator.getOutput(TableScanOperator.java:250)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:379)
	at com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:283)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:675)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:276)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1077)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:483)
	at com.facebook.presto.$gen.Presto_0_225_SNAPSHOT_2ec6bb0__testversion____20190820_080442_14390.run(Unknown Source)
	... 3 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: 3
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.populateOutputPositionsWithFilter(ListSelectiveStreamReader.java:428)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.readNotAllNulls(ListSelectiveStreamReader.java:382)
	at com.facebook.presto.orc.reader.ListSelectiveStreamReader.read(ListSelectiveStreamReader.java:227)
	at com.facebook.presto.orc.OrcSelectiveRecordReader.getNextPage(OrcSelectiveRecordReader.java:263)
	at com.facebook.presto.hive.orc.OrcSelectivePageSource.getNextPage(OrcSelectivePageSource.java:78)
	... 12 more
```

--- a/presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java
+++ b/presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java
@@ -29,6 +29,7 @@ import com.facebook.presto.spi.block.ArrayBlock;
 import com.facebook.presto.spi.block.Block;
 import com.facebook.presto.spi.block.BlockLease;
 import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import com.facebook.presto.spi.type.ArrayType;
 import com.facebook.presto.spi.type.Type;
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Maps;
@@ -71,7 +72,7 @@ public class ListSelectiveStreamReader
     private final boolean nonNullsAllowed;
     private final boolean outputRequired;
     @Nullable
-    private final Type outputType;
+    private final ArrayType outputType;
     // elementStreamReader is null if output is not required and filter is a simple IS [NOT] NULL
     @Nullable
     private final SelectiveStreamReader elementStreamReader;
@@ -129,7 +130,7 @@ public class ListSelectiveStreamReader
 
         this.streamDescriptor = requireNonNull(streamDescriptor, "streamDescriptor is null");
         this.outputRequired = requireNonNull(outputType, "outputType is null").isPresent();
-        this.outputType = outputType.orElse(null);
+        this.outputType = (ArrayType) outputType.orElse(null);
         this.level = subfieldLevel;
 
         if (listFilter != null) {
@@ -189,7 +190,7 @@ public class ListSelectiveStreamReader
     public int read(int offset, int[] positions, int positionCount)
             throws IOException
     {
-        checkArgument(positionCount > 0, "positionCount must be greater than zero");
+        checkArgument(positionCount > 0 || listFilter != null, "positionCount must be greater than zero");
         checkState(!valuesInUse, "BlockLease hasn't been closed yet");
 
         if (!rowGroupOpen) {
@@ -267,17 +268,25 @@ public class ListSelectiveStreamReader
                     if (nullsFilter.testNull()) {
                         outputPositionCount++;
                     }
+                    else {
+                        outputPositionCount -= nullsFilter.getPrecedingPositionsToFail();
+                        i += nullsFilter.getSucceedingPositionsToFail();
+                    }
                 }
             }
             else {
                 outputPositionCount = positionCount;
-                allNulls = true;
             }
         }
         else {
             outputPositionCount = 0;
         }
 
+        if (listFilter != null) {
+            listFilter.populateElementFilters(0, null, null, 0);
+        }
+
+        allNulls = true;
         return positions[positionCount - 1] + 1;
     }
 
@@ -371,6 +380,9 @@ public class ListSelectiveStreamReader
         }
 
         if (elementStreamReader != null && elementPositionCount > 0) {
+            elementStreamReader.read(elementReadOffset, elementPositions, elementPositionCount);
+        }
+        else if (listFilter != null && listFilter.getChild() != null) {
             elementStreamReader.read(elementReadOffset, elementPositions, elementPositionCount);
         }
         elementReadOffset += elementPositionCount + skippedElements;
@@ -470,11 +482,15 @@ public class ListSelectiveStreamReader
         boolean mayHaveNulls = nullsAllowed && presentStream != null;
 
         if (positionCount == outputPositionCount) {
+            Block elementBlock;
             if (elementOutputPositionCount == 0) {
-                return new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), outputPositionCount);
+                elementBlock = outputType.getElementType().createBlockBuilder(null, 0).build();
+            }
+            else {
+                elementBlock = elementStreamReader.getBlock(elementOutputPositions, elementOutputPositionCount);
             }
 
-            Block block = ArrayBlock.fromElementBlock(positionCount, Optional.ofNullable(mayHaveNulls ? nulls : null), offsets, elementStreamReader.getBlock(elementOutputPositions, elementOutputPositionCount));
+            Block block = ArrayBlock.fromElementBlock(positionCount, Optional.ofNullable(mayHaveNulls ? nulls : null), offsets, elementBlock);
             nulls = null;
             offsets = null;
             return block;
@@ -537,12 +553,15 @@ public class ListSelectiveStreamReader
             compactValues(positions, positionCount, includeNulls);
         }
 
+        BlockLease elementBlockLease;
         if (elementOutputPositionCount == 0) {
-            return newLease(new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), positionCount));
+            elementBlockLease = newLease(outputType.getElementType().createBlockBuilder(null, 0).build());
+        }
+        else {
+            elementBlockLease = elementStreamReader.getBlockView(elementOutputPositions, elementOutputPositionCount);
         }
 
         valuesInUse = true;
-        BlockLease elementBlockLease = elementStreamReader.getBlockView(elementOutputPositions, elementOutputPositionCount);
         Block block = ArrayBlock.fromElementBlock(positionCount, Optional.ofNullable(includeNulls ? nulls : null), offsets, elementBlockLease.get());
         return newLease(block, () -> closeBlockLease(elementBlockLease));
     }

--- a/presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java
+++ b/presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java
@@ -745,9 +745,12 @@ public class OrcTester
                 if (nestedValue == null) {
                     return filter == IS_NULL;
                 }
-                int index = toIntExact(((Subfield.LongSubscript) pathElement).getIndex());
+                int index = toIntExact(((Subfield.LongSubscript) pathElement).getIndex()) - 1;
                 nestedType = ((ArrayType) nestedType).getElementType();
-                nestedValue = ((List) nestedValue).get(index - 1);
+                if (index >= ((List) nestedValue).size()) {
+                    return true;
+                }
+                nestedValue = ((List) nestedValue).get(index);
             }
             else {
                 fail("Unsupported type: " + type);
@@ -808,6 +811,7 @@ public class OrcTester
         if (StandardTypes.ARRAY.equals(baseType)) {
             List<?> actualArray = (List<?>) actual;
             List<?> expectedArray = (List<?>) expected;
+            assertEquals(actualArray == null, expectedArray == null);
             assertEquals(actualArray.size(), expectedArray.size());
 
             Type elementType = type.getTypeParameters().get(0);

--- a/presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java
+++ b/presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java
@@ -18,6 +18,7 @@ import com.facebook.presto.orc.TupleDomainFilter.BigintValues;
 import com.facebook.presto.orc.TupleDomainFilter.BooleanValue;
 import com.facebook.presto.orc.TupleDomainFilter.DoubleRange;
 import com.facebook.presto.orc.TupleDomainFilter.FloatRange;
+import com.facebook.presto.spi.PrestoException;
 import com.facebook.presto.spi.Subfield;
 import com.facebook.presto.spi.type.SqlDate;
 import com.facebook.presto.spi.type.SqlTimestamp;
@@ -70,6 +71,8 @@ import static com.google.common.collect.Lists.newArrayList;
 import static java.util.Collections.nCopies;
 import static java.util.stream.Collectors.toList;
 import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
 
 public class TestSelectiveOrcReader
 {
@@ -371,12 +374,63 @@ public class TestSelectiveOrcReader
     }
 
     @Test
+    public void testArrayIndexOutOfBounds()
+            throws Exception
+    {
+        Random random = new Random(0);
+
+        // non-null arrays of varying sizes
+        try {
+            tester.testRoundTrip(arrayType(INTEGER),
+                    IntStream.range(0, 30_000).map(i -> random.nextInt(10)).mapToObj(size -> makeArray(size, random)).collect(toImmutableList()),
+                    ImmutableList.of(ImmutableMap.of(new Subfield("c[2]"), IS_NULL)));
+            fail("Expected 'Array subscript out of bounds' exception");
+        }
+        catch (PrestoException e) {
+            assertTrue(e.getMessage().contains("Array subscript out of bounds"));
+        }
+
+        // non-null nested arrays of varying sizes
+        try {
+            tester.testRoundTrip(arrayType(arrayType(INTEGER)),
+                    IntStream.range(0, 30_000).mapToObj(i -> ImmutableList.of(makeArray(random.nextInt(5), random), makeArray(random.nextInt(5), random))).collect(toImmutableList()),
+                    ImmutableList.of(ImmutableMap.of(new Subfield("c[2][3]"), IS_NULL)));
+            fail("Expected 'Array subscript out of bounds' exception");
+        }
+        catch (PrestoException e) {
+            assertTrue(e.getMessage().contains("Array subscript out of bounds"));
+        }
+
+        // empty arrays
+        try {
+            tester.testRoundTrip(arrayType(INTEGER),
+                    nCopies(30_000, ImmutableList.of()),
+                    ImmutableList.of(ImmutableMap.of(new Subfield("c[2]"), IS_NULL)));
+            fail("Expected 'Array subscript out of bounds' exception");
+        }
+        catch (PrestoException e) {
+            assertTrue(e.getMessage().contains("Array subscript out of bounds"));
+        }
+
+        // empty nested arrays
+        try {
+            tester.testRoundTrip(arrayType(arrayType(INTEGER)),
+                    nCopies(30_000, ImmutableList.of()),
+                    ImmutableList.of(ImmutableMap.of(new Subfield("c[2][3]"), IS_NULL)));
+            fail("Expected 'Array subscript out of bounds' exception");
+        }
+        catch (PrestoException e) {
+            assertTrue(e.getMessage().contains("Array subscript out of bounds"));
+        }
+    }
+
+    @Test
     public void testArraysOfNulls()
             throws Exception
     {
-        for (Type type : ImmutableList.of(BOOLEAN, BIGINT, INTEGER, SMALLINT, TINYINT, DOUBLE, REAL, TIMESTAMP)) {
+        for (Type type : ImmutableList.of(BOOLEAN, BIGINT, INTEGER, SMALLINT, TINYINT, DOUBLE, REAL, TIMESTAMP, arrayType(INTEGER))) {
             tester.testRoundTrip(arrayType(type),
-                    IntStream.range(0, 30_000).mapToObj(i -> Collections.nCopies(5, null)).collect(toImmutableList()),
+                    nCopies(30_000, nCopies(5, null)),
                     ImmutableList.of(
                             ImmutableMap.of(new Subfield("c[2]"), IS_NULL),
                             ImmutableMap.of(new Subfield("c[2]"), IS_NOT_NULL)));
@@ -418,7 +472,7 @@ public class TestSelectiveOrcReader
                 ImmutableList.of(
                         IntStream.range(0, 30_000).mapToObj(i -> random.nextInt()).collect(toImmutableList()),
                         IntStream.range(0, 30_000).boxed().map(i -> createMap(i)).collect(toImmutableList())),
-                ImmutableList.of(
+                toSubfieldFilters(
                         ImmutableMap.of(0, BigintRange.of(0, Integer.MAX_VALUE, false)),
                         ImmutableMap.of(1, IS_NOT_NULL),
                         ImmutableMap.of(1, IS_NULL)));
@@ -429,7 +483,7 @@ public class TestSelectiveOrcReader
                 ImmutableList.of(
                         IntStream.range(0, 30_000).mapToObj(i -> random.nextInt()).collect(toImmutableList()),
                         IntStream.range(0, 30_000).boxed().map(i -> i % 5 == 0 ? null : createMap(i)).collect(toList())),
-                ImmutableList.of(
+                toSubfieldFilters(
                         ImmutableMap.of(0, BigintRange.of(0, Integer.MAX_VALUE, false)),
                         ImmutableMap.of(1, IS_NOT_NULL),
                         ImmutableMap.of(1, IS_NULL),
@@ -442,7 +496,7 @@ public class TestSelectiveOrcReader
                 ImmutableList.of(
                         IntStream.range(0, 30_000).boxed().map(i -> i % 5 == 0 ? null : createMap(i)).collect(toList()),
                         IntStream.range(0, 30_000).mapToObj(i -> random.nextInt()).collect(toImmutableList())),
-                ImmutableList.of(
+                toSubfieldFilters(
                         ImmutableMap.of(0, IS_NULL, 1, BigintRange.of(0, Integer.MAX_VALUE, false)),
                         ImmutableMap.of(0, IS_NOT_NULL, 1, BigintRange.of(0, Integer.MAX_VALUE, false))));
     }

